hazelcast/src/main/java/com/hazelcast/instance/DefaultNodeExtension.java

```diff
@@ -236,7 +252,38 @@ public void onPartitionStateChange() {
     }
 
     @Override
+    public void onClusterVersionChange(Version newVersion) {
+        systemLogger.info("Cluster version changed to " + newVersion);
+        NodeEngineImpl nodeEngine = node.getNodeEngine();
+        ServiceManager serviceManager = nodeEngine.getServiceManager();
+        List<ClusterVersionListener> listeners = serviceManager.getServices(ClusterVersionListener.class);
+        for (ClusterVersionListener listener : listeners) {
+            listener.onClusterVersionChange(newVersion);
+        }
+        // also trigger cluster version change on explicitly registered listeners
+        for (ClusterVersionListener listener : clusterVersionListeners) {
+            listener.onClusterVersionChange(newVersion);
+        }
+    }
+
+    @Override
+    public boolean isNodeVersionCompatibleWith(Version clusterVersion) {
+        Preconditions.checkNotNull(clusterVersion);
+        return MAJOR_MINOR_VERSION_COMPARATOR.compare(node.getVersion(), clusterVersion) == 0;
+    }
+
+    @Override
     public boolean registerListener(Object listener) {
+        if (listener instanceof HazelcastInstanceAware) {
+            ((HazelcastInstanceAware) listener).setHazelcastInstance(node.hazelcastInstance);
+        }
+        if (listener instanceof ClusterVersionListener) {
+            ClusterVersionListener clusterVersionListener = (ClusterVersionListener) listener;
+            clusterVersionListeners.add(clusterVersionListener);
+            // on registration, invoke once the listening method so version is properly initialized on the listener
+            clusterVersionListener.onClusterVersionChange(getClusterOrNodeVersion());
```

Comment:
There are listener tests in `DefaultNodeExtensionTest` which verify the listener is invoked on registration.